Generate RPM and support arbitrary files copied to the host

giuseppe opened the pull request:
this is a prototype to add "pre installed" system containers.  The idea is to have another deployment tree under `/usr/lib/containers/atomic` that is not handled by "atomic install/uninstall/update", where '/var/lib/containers/atomic' is fully handled by atomic.  The systemd files are also copied under '/usr/lib'.

The second patch adds the possibility to install a system container into a rpm.  The `---generate-rpm` (three dashes) option is hidden for now.   The spec file is auto generated, and stuff like `Requires`, `Provides`, `Conflicts` can be specified through labels.  So in the Dockerfile something like: `LABEL Provides="etcd"`, will be translated directly to the specfile.

Example:
```
# atomic --debug install --system ---generate-rpm gscrivano/etcd  
# rpm -i x86_64/atomic-container-etcd-1.0-1.0.x86_64.rpm
```

jfilak commented on Atomic/syscontainers.py:
```diff
@@ -1355,3 +1361,77 @@ def get_out_checksum(obj): return obj.out_checksum if hasattr(obj, 'out_checksum
+        for root, _, files in os.walk(os.path.join(destdir, "etc")):
+            rel_path = os.path.relpath(root, destdir)
+            for f in files:
+                spec += "%config \"%s\"\n" % os.path.join("/", rel_path, f)
```
I'm confused, this line looks like you want to install container's /etc to host's /etc.

giuseppe replied:
here I am mapping from the temporary directory (where the container was checked out) to the absolute path `/etc' in the host (where the .rpm will install these files)

jfilak replied:
Oh sorry, I thought that the destdir variable points to the container's root.

rhatdan commented on Atomic/install.py:
```diff
@@ -59,6 +59,9 @@ def cli(subparser):
                 system_xor_user.add_argument("--system", dest="system",
                                              action='store_true', default=False,
                                              help=_('install a system container'))
+                installp.add_argument("---generate-rpm", dest="generate_rpm",
```
I would prefer that the user not need to specify this.  This should happen automatically if the image supports generation of rpms for install.

rhatdan replied:
Users will forget to do this, and we want to eventually make this the default for all of our system containers.

giuseppe replied:
the --generate-rpm doesn't probably belong here, I just  didn't know where to add it.

I think we will need a way to just generate the .rpm, so that it can be used as any regular .rpm, you can even install it without using atomic at all.

Installing the .rpm could be the default here, but we will need a way to specify also to install the container in the usual way using `/var/lib` instead of '/usr/lib' if needed.

rhatdan replied:
Right I would say that is the fallback.  
I believe the user puts the data in as templates rather then in labels.  I am thinking a rpm.yaml file which specifies information for the spec file, but allows atomic CLI to do substitutions like the name of the container so the spec file ends up with foobar_etcd, and we could prefix all executables or paths with the container names.

rhatdan commented on Atomic/syscontainers.py:
```diff
@@ -419,7 +432,11 @@ def _do_checkout(self, repo, name, img, upgrade, values, destination, unitfileou
 
-        values["DESTDIR"] = destination
+        if prefix:
+            values["DESTDIR"] = os.path.join("/", os.path.relpath(destination, prefix))
```
values["DESTDIR"] = (os.path.join("/", os.path.relpath(destination), prefix)) if prefix else destination)

rhatdan commented on Atomic/syscontainers.py:
```diff
@@ -501,24 +518,28 @@ def _write_template(inputfilename, data, values, destination):
-        if os.path.exists(sym):
-            os.unlink(sym)
-        os.symlink(destination, sym)
+        if not prefix:
```
Should we just
```
if prefix:
     return
```

rhatdan commented on Atomic/syscontainers.py:
```diff
@@ -1317,3 +1361,77 @@ def get_out_checksum(obj): return obj.out_checksum if hasattr(obj, 'out_checksum
+                   "--define", "_rpmdir %s" % cwd,
+                   "--build-in-place",
+                   "--buildroot=%s" % rpm_content]
+            util.write_out(" ".join(cmd))
```
if self.debug:

rhatdan commented:
I think there should be a rpm specification template in the container image that triggers the rpm build for install.  This should be simpler then the rpm.spec,   Perhaps just the files section as well as description, license, summary and potentially some other fields.

Should we default post install of rpm package to 

atomic uninstall CONTAINER

giuseppe commented:
We can add a spec template, as we do with the other configuration files (and offer a substitution for the list of the installed files, it would be very tedious to list them otherwise).  I thought of using labels so we will be able to query resolves/provides/conflicts remotely without fetching the image first.

I have not added a preun script yet.  I think it should just be to stop the systemd service.  I think "atomic uninstall" should not interfere with rpm to handle/delete the files.

rhatdan commented:
Right I guess we could get a circular dependency here.  
I guess atomic uninstall should execute

rpm -e PACKAGE.

giuseppe commented:
@rhatdan I have addressed your comments for the code.

I have pushed a bunch of other patches to cleanly support copied files to the host.

I have included the idea of @jfilak to copy files from `/exports/hostfs` to the host.  These files are tracked by the rpm when generating one, otherwise they are managed by atomic and removed on "atomic uninstall".

The files copied to the host can be also template files, the same substitution rules that are used for the runc configuration file and the systemd service file apply here.

It is possible to specify that a system container doesn't install any service, so that can be used just as a way to copy files to the host, or to generate an rpm without the overhead of the service+rootfs files.

rhatdan commented:
@TomasTomecek PTAL

rhatdan commented:
@baude PTAL

jfilak commented:
I like this pull request and I told @giuseppe that he has done a great job in private chat. However, there is one thing that concerns me a bit. I think it's a pity that the whole system container thing is bundled in 'atomic' utility. I develop my containers on Fedora and I would appreciate a tool capable of installing a test container on my development machine.

rhatdan commented:
atomic command is available on Fedora as well as atomic host.  It is a separate package.

cgwalters commented:
Atomic Host is part of Fedora.  It *is* Fedora as much as a system managed via yum.

rhatdan commented:
Right  atomic CLI is available on Fedora Server, Fedora Workstation, Fedora Cloud/Atomic

jfilak commented:
So it is possible to develop a DNF plugin installing system containers through 'atomic' utility. It would be cool to run:
```
dnf container-install gscrivano/etcd
```

TomasTomecek commented on Atomic/syscontainers.py:
```diff
@@ -1317,3 +1493,86 @@ def get_out_checksum(obj): return obj.out_checksum if hasattr(obj, 'out_checksum
+        if installed_files:
+            for i in installed_files:
+                spec = spec + "%s\n" % i
+
```
Would it make sense to add changelog? So users know that the rpm is generated and installed by `atomic`.

TomasTomecek commented:
Got this traceback with image you provided to me:

```
$ sudo ./atomic --debug install --system ---generate-rpm --display img
Extracting to /tmp/tmprTqPaB/rpmroot/usr/lib/containers/atomic/img
./atomic --debug install --system ---generate-rpm --display img: too few arguments
Try './atomic --debug install --system ---generate-rpm --display img --help' for more information.
Traceback (most recent call last):
  File "./atomic", line 187, in <module>
    sys.exit(_func())
  File "/home/tt/g/atomic/Atomic/install.py", line 90, in install
    return self.syscontainers.install(self.image, self.name)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 169, in install
    repo = self._get_ostree_repo()
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 1507, in generate_rpm
    self._checkout(repo, name, image, 0, False, destination=rootfs, prefix=rpm_content)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 276, in _checkout
    return self._do_checkout(repo, name, img, upgrade, values, destination, unitfileout, tmpfilesout, extract_only, remote, prefix, installed_files
=installed_files)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 363, in _do_checkout
    was_service_active = self._is_service_active(name)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 965, in _is_service_active
    return self._systemctl_command("is-active", name, quiet=True).replace("\n", "") == "active"
AttributeError: 'NoneType' object has no attribute 'replace'
```

giuseppe commented:
@TomasTomecek  `--display` is not really supported when generating an rpm as we need to read information from the files that are installed/checked out from the OSTree repository.  I have added a fixup patch though to not cause the traceback when `--display` is used.

rhatdan commented:
I would like to setup a BlueJeans tomorrow to help design this tool to make sure everyone is on the same page.  We need this to work for use cases outside of --system containers as well.

TomasTomecek commented:
> We need this to work for use cases outside of --system containers as well.

This would be awesome. Right now I'm in a process of writing down use cases which could benefit greatly from this feature. Please, invite me to the discussion.

giuseppe commented:
yes, this PR makes possible to install arbitrary files on the host and track them with a rpm without requiring a running or even installed container/systemd service.

Every file under /exports/hostfs will be copied to the host file system.  If `noContainerService` is used in the `manifest.json` file then atomic won't try to install the systemd service and the runc container.

rhatdan commented:
@giuseppe I scheduled the meeting tomorrow at 8:30 AM EST.  I would like you to take the lead to describe what you are building and display different parts.  Could you throw together a Google Doc in the mean time to help facilitate the discussion.

rhatdan commented:
BTW Add anyone you would like to the meeting.

rhatdan commented:
@TomasTomecek If you could write up your Use Cases, that would be helpful also.

giuseppe commented:
I've added a new patch that allows the user to directly specify a spec file, with these rules:

- If `/exports/rpm.spec` is present, then it is used as it is, otherwise,
- If `/exports/rpm.spec.template` is present, it is first preprocessed as any template file and then it is used, otherwise
- We generate a spec file

rhatdan commented:
@giuseppe @baude we need to start moving forward on this.

giuseppe commented:
@rhatdan we can probably skip the "automatically install the rpm if .spec is defined" part for now.  We can add this later once we are able to install local rpm on Atomic Host.  We will still need to be able to build rpm offline anyway.  Not sure where the option should be placed though. "atomic install --system --only-generate-rpm" ?

rhatdan commented:
I am on PTO this week. I guess you and @baude need to figure this out.  We can discuss the options next Monday, but that should not prevent you guys from hacking on it.

giuseppe commented:
@rhatdan let's just decide how to present the offline "generate rpm" functionality, which probably we want to have anyway.  We can postpone the automatic installation of an .rpm later, once we can do it on Atomic Host (and also have clearer ideas on when/how to use it).

rhatdan commented:
I would prefer to have this be automatic and not involve the user at all.  The atomic command and image should have enough data that atomic CLI can figure out if it needs to build an RPM.

rhatdan commented:
@giuseppe should we have a meeting to discuss this feature?  Or do you have enough to start?

giuseppe commented:
@rhatdan we can setup a meeting so we can discuss the last details.

@rhatdan I think we will still need to be able to build an rpm offline though, so that it can be included in the Atomic Host image if needed.  @cgwalters do you think we should allow this?

TomSweeneyRedHat commented:
Hi Giuseppe,

If possible, could you include me on the meeting invite too please?  I'd
like to sit in for informational purposes.

Thanks!

t

On 03/02/2017 10:10 AM, Giuseppe Scrivano wrote:
>
> @rhatdan <https://github.com/rhatdan> we can setup a meeting so we can
> discuss the last details.
>
> @rhatdan <https://github.com/rhatdan> I think we will still need to be
> able to build an rpm offline though, so that it can be included in the
> Atomic Host image if needed. @cgwalters <https://github.com/cgwalters>
> do you think we should allow this?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/projectatomic/atomic/pull/767#issuecomment-283679588>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYbWBQwzH3SU5GNvmesI27VEy3X5U9Mlks5rhtv8gaJpZM4K6etD>.
>

TomasTomecek commented:
I would also appreciate invite. @giuseppe will you be in Westford next week?

giuseppe commented:
@TomasTomecek no, I won't be in Westford.  Sure I'll send an invitation to you for the meeting

rhatdan commented:
I added @baude @cgwalters @sctweedie and Stef Walters to invite also

giuseppe commented:
While there is probably more to do (e.g. show some information in `atomic show`) and more tests, I think it is fine to give it a try.  I dropped the [RFC][WIP] tags to make it clearer.

rhatdan commented:
@baude PTAL

TomasTomecek commented:
I played with this today and wasn't able to make it work, unfortunately.

This is the exception I got:

```
$ sudo ./atomic --debug install --system --system-package=build docker:modularitycontainers/nginx
Namespace(_class=<class 'Atomic.install.Install'>, args=[], assumeyes=False, debug=True, display=False, func='install', image='docker:modularitycontainers/nginx', name=None,
opt1=None, opt2=None, opt3=None, remote=None, setvalues=None, storage=None, system=True, system_package='build', user=False)
g-io-error-quark: Not a regular file or symlink: loop1 (0)
Traceback (most recent call last):
  File "./atomic", line 189, in <module>
    sys.exit(_func())
  File "/home/tt/g/atomic/Atomic/install.py", line 106, in install
    return self.syscontainers.install(self.image, self.name)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 215, in install
    image = self._pull_image_to_ostree(repo, image, False)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 160, in _pull_image_to_ostree
    image = self._pull_docker_image(repo, image.replace("docker:", "", 1))
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 1401, in _pull_docker_image
    return self._pull_docker_tar(repo, temptar.name, image)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 1422, in _pull_docker_tar
    self._pull_dockertar_layers(repo, imagebranch, temp_dir, input_layers, labels=labels)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 1595, in _pull_dockertar_layers
    self._import_layers_into_ostree(repo, imagebranch, manifest, layers_to_import)
  File "/home/tt/g/atomic/Atomic/syscontainers.py", line 1367, in _import_layers_into_ostree
    repo.write_directory_to_mtree(Gio.File.new_for_path(temp_dir), mtree, modifier)
Error: g-io-error-quark: Not a regular file or symlink: loop1 (0)
```

[This is](https://github.com/container-images/nginx/blob/64c2f087c19f6fb20ab33e8a1ed96e5a5a4e17ab/Dockerfile) my Dockerfile.

What's really interesting is that when I checked files in the temp directory (`/var/lib/containers/atomic/.Bze55J`), the `/dev` was fully populated and that's what I'm assuming caused the error above:

```
/var/lib/containers/atomic/.Bze55J $ ll dev
total 0
crw------- 1 root mock  5,   1 Mar  8 19:10 console
lrwxrwxrwx 1 root mock      13 Mar 20 17:28 fd -> /proc/self/fd
crw-rw-rw- 1 root mock  1,   7 Mar  8 19:10 full
brw-rw-rw- 1 root mock  7,   0 Mar  8 19:10 loop0
brw-rw-rw- 1 root mock  7,   1 Mar  8 19:10 loop1
brw-rw-rw- 1 root mock  7,   2 Mar  8 19:10 loop2
brw-rw-rw- 1 root mock  7,   3 Mar  8 19:10 loop3
brw-rw-rw- 1 root mock  7,   4 Mar  8 19:10 loop4
crw-rw-rw- 1 root mock 10, 237 Mar  8 19:10 loop-control
crw-rw-rw- 1 root mock  1,   3 Mar  8 19:10 null
lrwxrwxrwx 1 root mock       8 Mar 20 17:28 ptmx -> pts/ptmx
drwxrwsr-x 2 root mock       6 Mar  8 19:10 pts
crw-rw-rw- 1 root mock  1,   8 Mar  8 19:10 random
drwxrwsr-x 2 root mock       6 Mar  8 19:10 shm
lrwxrwxrwx 1 root mock      15 Mar 20 17:28 stderr -> /proc/self/fd/2
lrwxrwxrwx 1 root mock      15 Mar 20 17:28 stdin -> /proc/self/fd/0
lrwxrwxrwx 1 root mock      15 Mar 20 17:28 stdout -> /proc/self/fd/1
crw-rw-rw- 1 root tty   5,   0 Mar  8 19:10 tty
cr--r--r-- 1 root mock  1,   9 Mar  8 19:10 urandom
crw-rw-rw- 1 root mock  1,   5 Mar  8 19:10 zero
```

Also, we would really appreciate if we could use this functionality, copying files to host & creating an RPM out of them, with docker containers.

giuseppe commented:
sorry I probably gave you the wrong info.  To pull an image to the OSTree storage you need to do:

`./atomic pull --storage ostree docker:modularitycontainers/nginx`

and once it is stored in OSTree then you can install it:

`./atomic install --system modularitycontainers/nginx`

We could probably make atomic install smarter and do the previous steps at once.

I am not sure how this could be done with Docker containers, as differently than system containers, it is not enough to copy some files to get the container seen by Docker.  If you refer to copy some files to the host part, we could probably do that, just you'll have to bind the .rpm to the docker container lifecycle somehow (i.e. you can uninstall the docker container but the .rpm is still installed).  Do you have any example of what kind of files do you need to copy from a Docker container?

rhatdan commented:
Yes 
atomic install --system docker:modularitycontainers/nginx  
Should pull the image and store it in OSTree.

giuseppe commented:
@rhatdan I have done few more changes.  A lot of changes are accumulated here and if there are no objections, I think it is fine to merge, some parts need to be cleaned up and also tests need to be added but we can do it on master.
I know that @ashcrow is working on the same portion on the code, getting this merged will probably save some merge conflicts.

rhatdan commented:
Ok I would prefer to wait to merge until tomorrow, so we can get RHEL release out, without this patch.

giuseppe commented:
@rhatdan yes sure, let's give this some time before it gets into a release

rhatdan commented:
@rh-atomic-bot r+ 4321643
